### PR TITLE
Use masquerade default gateway when no default gw is found

### PR DESCRIPTION
**Pull request by trozet**
When no default gateway is found, OVNK will still come up. This behavior was intentional: 3eeb930791e04824b76b03172f64a1db21d9c377

The goal was that with design changes, host->service traffic would work without the need for a default gateway to be detected. This works, in most cases, but there is a case we did not consider. If a user has deployed a cluster with local gateway mode, and is using a different interface/network for kapi traffic (not the gateway bridge interface), then endpoints to the kubernetes service will reside on this other network. In this case when a host tries to talk to kube API service, the traffic would go to the GR, it would be DNAT'ed to an IP address not on any known network, and be dropped by OVN routing when there is no default gateway route.

When the configuration parameter AllowNoUplink is set, we will set the default route to be the masquerade IP, which solves this problem. This commit changes the behavior to not require AllowNoUplink to be set in order to achieve the same behavior, as services should work in this scenario even without a default gw.

### CODE REVIEW

**Review by tssurya — COMMENTED** [review]
overall lgtm just a question around the warning print
also wider question, I wasn't involved in the design of `AllowNoUplink` option where we deploy LGW without interfacing with host NIC, but currently this is really used only in `bridgeForInterface` anyways just checking this option is still relevant in OVNK considering we decided to add the default route irrespective of whether this option is set or not

**Comment by trozet** [comment]
> also wider question, I wasn't involved in the design of `AllowNoUplink` option where we deploy LGW without interfacing with host NIC, but currently this is really used only in `bridgeForInterface` anyways just checking this option is still relevant in OVNK considering we decided to add the default route irrespective of whether this option is set or not

good question. So I debated removing allowNoUplink as well. I think the one key difference is this option still allows for no uplink at all to be found/attached to the bridge:
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/gateway.go#L494C3-L494C15

So without this option, if we cannot find an interface to attach to the bridge, then we consider it as a failed state unless explicitly told with allowNoUplink. I'm not sure how useful this is or not, but I decided to leave it alone since it isnt broke :)

File: go-controller/pkg/node/gateway_init.go [hunk 1]
```diff
@@ -184,10 +184,15 @@ func getGatewayNextHops() ([]net.IP, string, error) {
 			}
 			gatewayIntf = defaultGatewayIntf
 		} else {
-			if gatewayIntf != defaultGatewayIntf || len(defaultGatewayNextHops) == 0 {
-				if config.Gateway.Mode == config.GatewayModeLocal && config.Gateway.AllowNoUplink {
+			if gatewayIntf != defaultGatewayIntf {
```
Comment (trozet) [inline go-controller/pkg/node/gateway_init.go:187]: looking at this a second time... I think this code is not 100% correct. If we find the defaultGatewayIntf does not match the configured gateway, we should not be using the defaultGatewayNextHops for that interface in OVN. So I should fix this to match the previous if statement logic of:

if gatewayIntf != defaultGatewayIntf || len(defaultGatewayNextHops) == 0 {

wdyt @tssurya ?

File: go-controller/pkg/node/gateway_init.go [hunk 2]
```diff
@@ -184,10 +184,15 @@ func getGatewayNextHops() ([]net.IP, string, error) {
...
+				klog.Warningf("Found default gateway interface: %q does not match provided interface from config: %q", defaultGatewayIntf, gatewayIntf)
+			}
+			if len(defaultGatewayNextHops) == 0 {
+				klog.Warning("No default route identified in the host. Egress features and pod egress traffic in " +
+					"shared gateway mode may not function correctly!")
```
[the remark this answers is not in the document]
Comment (trozet) [reply go-controller/pkg/node/gateway_init.go:192]: I agree for allowNoUplink, we can probably skip printing it. 

I guess I am also missing a comma. It should be:
Egress features, and pod egress traffic in shared gateway mode may not function correctly

Which I meant to say only egress pod traffic in shared gateway mode might not work correctly. Egress pod traffic in LGW will still work. Egress features may not work in both modes.

Should I reword it to clarify they are 2 separate things?

File: go-controller/pkg/node/gateway_init.go (hunk 1 above)
Comment (tssurya) [reply go-controller/pkg/node/gateway_init.go:187]: so... its more like if the default gateway did match configured gateway then we won't enter this if but we should enter because what if default gateway nexthops is 0 right?
as in you converted the || into && which changed the logic from prior behaviour.
TBH: the if block nesting was/is a mess here :( if you could clear the clutter up.. :D

Comment (trozet) [reply go-controller/pkg/node/gateway_init.go:187]: It wasnt turned into && it just doesnt matter in the current code if the interfaces dont match. Let me push a new revision and I think you will see what I mean.

Comment (tssurya) [reply go-controller/pkg/node/gateway_init.go:187]: ack I'll wait for the push, my point was:
before your change it was:
```
if gatewayIntf != defaultGatewayIntf || len(defaultGatewayNextHops) == 0 {
    if config.Gateway.Mode == config.GatewayModeLocal {
```
and after your change it became:
```
if gatewayIntf != defaultGatewayIntf {
}
if len(defaultGatewayNextHops) == 0 {
    if config.Gateway.Mode == config.GatewayModeLocal {
    }
}
```
which is not && sorry :) that was my confusion
and right, so the total logic needs to be either we have 0 nexthops OR we have a mismatch of interfaces versus what we have currently in your code is, if its an interface mismatch we don't do anything except logging the warning and doing the default gateway plumbing anyways which is wrong as we should be plumbing in the masquerade bits.
ok we are on same page.

**Comment by tssurya** [comment]
@trozet : wanna repush here?

**Comment by trozet** [comment]
addressed the comments in last push, will now push another with just rebase

File: go-controller/pkg/node/gateway_init.go [hunk 3]
```diff
@@ -184,10 +184,19 @@ func getGatewayNextHops() ([]net.IP, string, error) {
 			}
 			gatewayIntf = defaultGatewayIntf
 		} else {
+			if gatewayIntf != defaultGatewayIntf {
```
Comment (tssurya) [inline go-controller/pkg/node/gateway_init.go:187]: too much if's and else's for my taste :) but we can clean code later, I see value in printing different warnings, will help us debug better in the future.

**Review by tssurya — APPROVED** [review]

File: go-controller/pkg/node/gateway_init_linux_test.go [hunk 4]
```diff
@@ -1857,7 +1857,6 @@ var _ = Describe("Gateway unit tests", func() {
 				gwIPs := []net.IP{config.Gateway.MasqueradeIPs.V4DummyNextHopMasqueradeIP}
 				config.Gateway.Interface = dummyBridgeName
 				config.Gateway.Mode = config.GatewayModeLocal
-				config.Gateway.AllowNoUplink = true
```
Comment (tssurya) [inline go-controller/pkg/node/gateway_init_linux_test.go:1860]: self-note: there is another UT testing with allowNoUplink to true so we do have coverage.
